// mongodb-stitch-core-sdk/dist/umd/auth/internal/models/StoreAuthInfo.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../AuthInfo", "./StoreCoreUserProfile", "./StoreStitchUserIdentity"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var AuthInfo_1 = __importDefault(require("../AuthInfo"));
    var StoreCoreUserProfile_1 = __importDefault(require("./StoreCoreUserProfile"));
    var StoreStitchUserIdentity_1 = __importDefault(require("./StoreStitchUserIdentity"));
    var Fields;
    (function (Fields) {
        Fields["USER_ID"] = "user_id";
        Fields["DEVICE_ID"] = "device_id";
        Fields["ACCESS_TOKEN"] = "access_token";
        Fields["REFRESH_TOKEN"] = "refresh_token";
        Fields["LOGGED_IN_PROVIDER_TYPE"] = "logged_in_provider_type";
        Fields["LOGGED_IN_PROVIDER_NAME"] = "logged_in_provider_name";
        Fields["USER_PROFILE"] = "user_profile";
    })(Fields || (Fields = {}));
    function readFromStorage(storage) {
        var rawInfo = storage.get(StoreAuthInfo.STORAGE_NAME);
        if (!rawInfo) {
            return undefined;
        }
        return StoreAuthInfo.decode(JSON.parse(rawInfo));
    }
    exports.readFromStorage = readFromStorage;
    function writeToStorage(authInfo, storage) {
        var info = new StoreAuthInfo(authInfo.userId, authInfo.deviceId, authInfo.accessToken, authInfo.refreshToken, authInfo.loggedInProviderType, authInfo.loggedInProviderName, authInfo.userProfile
            ? new StoreCoreUserProfile_1.default(authInfo.userProfile.userType, authInfo.userProfile.data, authInfo.userProfile.identities.map(function (identity) {
                return new StoreStitchUserIdentity_1.default(identity.id, identity.providerType);
            }))
            : undefined);
        storage.set(StoreAuthInfo.STORAGE_NAME, JSON.stringify(info.encode()));
    }
    exports.writeToStorage = writeToStorage;
    var StoreAuthInfo = (function (_super) {
        __extends(StoreAuthInfo, _super);
        function StoreAuthInfo(userId, deviceId, accessToken, refreshToken, loggedInProviderType, loggedInProviderName, userProfile) {
            var _this = _super.call(this, userId, deviceId, accessToken, refreshToken, loggedInProviderType, loggedInProviderName, userProfile) || this;
            _this.userProfile = userProfile;
            return _this;
        }
        StoreAuthInfo.decode = function (from) {
            var userId = from[Fields.USER_ID];
            var deviceId = from[Fields.DEVICE_ID];
            var accessToken = from[Fields.ACCESS_TOKEN];
            var refreshToken = from[Fields.REFRESH_TOKEN];
            var loggedInProviderType = from[Fields.LOGGED_IN_PROVIDER_TYPE];
            var loggedInProviderName = from[Fields.LOGGED_IN_PROVIDER_NAME];
            var userProfile = from[Fields.USER_PROFILE];
            return new StoreAuthInfo(userId, deviceId, accessToken, refreshToken, loggedInProviderType, loggedInProviderName, StoreCoreUserProfile_1.default.decode(userProfile));
        };
        StoreAuthInfo.prototype.encode = function () {
            var to = {};
            to[Fields.USER_ID] = this.userId;
            to[Fields.ACCESS_TOKEN] = this.accessToken;
            to[Fields.REFRESH_TOKEN] = this.refreshToken;
            to[Fields.DEVICE_ID] = this.deviceId;
            to[Fields.LOGGED_IN_PROVIDER_NAME] = this.loggedInProviderName;
            to[Fields.LOGGED_IN_PROVIDER_TYPE] = this.loggedInProviderType;
            to[Fields.USER_PROFILE] = this.userProfile
                ? this.userProfile.encode()
                : undefined;
            return to;
        };
        StoreAuthInfo.STORAGE_NAME = "auth_info";
        return StoreAuthInfo;
    }(AuthInfo_1.default));
    exports.StoreAuthInfo = StoreAuthInfo;
});
