// mongodb-stitch-core-sdk/dist/umd/auth/internal/AuthInfo.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var AuthInfo = (function () {
        function AuthInfo(userId, deviceId, accessToken, refreshToken, loggedInProviderType, loggedInProviderName, userProfile) {
            this.userId = userId;
            this.deviceId = deviceId;
            this.accessToken = accessToken;
            this.refreshToken = refreshToken;
            this.loggedInProviderType = loggedInProviderType;
            this.loggedInProviderName = loggedInProviderName;
            this.userProfile = userProfile;
        }
        AuthInfo.empty = function () {
            return new AuthInfo(undefined, undefined, undefined, undefined, undefined, undefined, undefined);
        };
        AuthInfo.prototype.loggedOut = function () {
            return new AuthInfo(undefined, this.deviceId, undefined, undefined, undefined, undefined, undefined);
        };
        AuthInfo.prototype.merge = function (newInfo) {
            return new AuthInfo(newInfo.userId === undefined ? this.userId : newInfo.userId, newInfo.deviceId === undefined ? this.deviceId : newInfo.deviceId, newInfo.accessToken === undefined
                ? this.accessToken
                : newInfo.accessToken, newInfo.refreshToken === undefined
                ? this.refreshToken
                : newInfo.refreshToken, newInfo.loggedInProviderType === undefined
                ? this.loggedInProviderType
                : newInfo.loggedInProviderType, newInfo.loggedInProviderName === undefined
                ? this.loggedInProviderName
                : newInfo.loggedInProviderName, newInfo.userProfile === undefined ? this.userProfile : newInfo.userProfile);
        };
        return AuthInfo;
    }());
    exports.default = AuthInfo;
});
