// mongodb-stitch-core-services-mongodb-remote/dist/umd/internal/ResultDecoders.js
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../RemoteInsertManyResult"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var RemoteInsertManyResult_1 = __importDefault(require("../RemoteInsertManyResult"));
    var RemoteInsertManyResultFields;
    (function (RemoteInsertManyResultFields) {
        RemoteInsertManyResultFields["InsertedIds"] = "insertedIds";
    })(RemoteInsertManyResultFields || (RemoteInsertManyResultFields = {}));
    var RemoteInsertOneResultFields;
    (function (RemoteInsertOneResultFields) {
        RemoteInsertOneResultFields["InsertedId"] = "insertedId";
    })(RemoteInsertOneResultFields || (RemoteInsertOneResultFields = {}));
    var RemoteUpdateResultFields;
    (function (RemoteUpdateResultFields) {
        RemoteUpdateResultFields["MatchedCount"] = "matchedCount";
        RemoteUpdateResultFields["ModifiedCount"] = "modifiedCount";
        RemoteUpdateResultFields["UpsertedId"] = "upsertedId";
    })(RemoteUpdateResultFields || (RemoteUpdateResultFields = {}));
    var RemoteDeleteResultFields;
    (function (RemoteDeleteResultFields) {
        RemoteDeleteResultFields["DeletedCount"] = "deletedCount";
    })(RemoteDeleteResultFields || (RemoteDeleteResultFields = {}));
    var RemoteInsertManyResultDecoder = (function () {
        function RemoteInsertManyResultDecoder() {
        }
        RemoteInsertManyResultDecoder.prototype.decode = function (from) {
            return new RemoteInsertManyResult_1.default(from[RemoteInsertManyResultFields.InsertedIds]);
        };
        return RemoteInsertManyResultDecoder;
    }());
    var RemoteInsertOneResultDecoder = (function () {
        function RemoteInsertOneResultDecoder() {
        }
        RemoteInsertOneResultDecoder.prototype.decode = function (from) {
            return {
                insertedId: from[RemoteInsertOneResultFields.InsertedId]
            };
        };
        return RemoteInsertOneResultDecoder;
    }());
    var RemoteUpdateResultDecoder = (function () {
        function RemoteUpdateResultDecoder() {
        }
        RemoteUpdateResultDecoder.prototype.decode = function (from) {
            return {
                matchedCount: from[RemoteUpdateResultFields.MatchedCount],
                modifiedCount: from[RemoteUpdateResultFields.ModifiedCount],
                upsertedId: from[RemoteUpdateResultFields.UpsertedId]
            };
        };
        return RemoteUpdateResultDecoder;
    }());
    var RemoteDeleteResultDecoder = (function () {
        function RemoteDeleteResultDecoder() {
        }
        RemoteDeleteResultDecoder.prototype.decode = function (from) {
            return {
                deletedCount: from[RemoteDeleteResultFields.DeletedCount]
            };
        };
        return RemoteDeleteResultDecoder;
    }());
    var ResultDecoders = (function () {
        function ResultDecoders() {
        }
        ResultDecoders.remoteInsertManyResultDecoder = new RemoteInsertManyResultDecoder();
        ResultDecoders.remoteInsertOneResultDecoder = new RemoteInsertOneResultDecoder();
        ResultDecoders.remoteUpdateResultDecoder = new RemoteUpdateResultDecoder();
        ResultDecoders.remoteDeleteResultDecoder = new RemoteDeleteResultDecoder();
        return ResultDecoders;
    }());
    exports.default = ResultDecoders;
});
